src/box/iproto.cc

```diff
@@ -2759,6 +2793,36 @@ iproto_session_notify(struct session *session,
 
 /** }}} */
 
+static void
+iproto_send_stop_msg(void);
+
+/**
+ * Shutdown trigger callback.
+ *  1. Stops accepting new connections.
+ *  2. Sends an IPROTO_SHUTDOWN packet to all clients that support
+ *     the graceful shutdown feature.
+ *  3. Waits for the clients to close connections.
+ */
+static int
+iproto_on_shutdown_f(void *arg)
+{
+	(void)arg;
+	iproto_send_stop_msg();
```

Comment:
this is not enough to guarantee that no connection will be established after the shutdown starts. At the time of calling this function, a connection can be established in the iproto thread. This connection will be added to `shutdown_list` when this function sleep on `shutdown_list_empty_cond`. I think we should also add some global flag `shutdown_is_started`, which we should set to true in this function. We should check this flag in `tx_process_connect` and return error to client if shutdown is started.

Reply:
Right. Thanks for catching this! I added a flag so that now if `IPROTO_ID` is received after the server started shutting down, it will send `IPROTO_SHUTDOWN` packet along with a reply to `IPROTO_ID`. I also patched net.box so that if it receives `IPROTO_SHUTDOWN` before it starts serving requests (during handshake or auth), it closes the connection right away. I added a couple of test cases to `box-luatest/graceful_shutdown_errinj.test.lua` to check this behavior.